hanish/exceptions.py

```diff
@@ -0,0 +1,66 @@
+# hanish.exceptions
+# Exceptions hierarchy for the hanish application.
+#
+# Author:   Benjamin Bengfort <benjamin@bengfort.com>
+# Created:  Mon May 08 12:44:41 2017 -0400
+#
+# Copyright (C) 2016 Bengfort.com
+# For license information, see LICENSE.txt
+#
+# ID: exceptions.py [] benjamin@bengfort.com $
+
+"""
+Exceptions hierarchy for the hanish application.
+"""
+
+##########################################################################
+## Exception Hierarchy
+##########################################################################
+
+class HanishException(Exception):
```

Comment:
I'm a big fan of exception programming, especially since I started in C and currently do a lot of Go programming and neither of those languages have an exception feature. As a data scientist, I know that errors propagate exponentially through the modeling stack, and there is nothing worse than an undetected data error. As a programmer, I like the flexibility to handle behavior at any level of the stack. 

So subclass hierarchy vs. messages. In Python, the `try/except` block allows you to filter exception handling by type. So for example, this is very common in my code:

```python
try: 
    # do something 
except (InputError, NotFoundError):
    # alert the user to try again 
except (DatabaseError, APIError):
    # log the error and tell the user to contact the admin 
except AppException:
    # email the admins that something unhandled happened 
except Exception:
    # panic gracefully 
```

Note that `except` filters everything that is a subclass of the specified exception, so that last `except Exception` catches _all_ exceptions, the second to last captures all exceptions we defined, and then there is handling for other exceptions. 

You could also test the string in the message, but I tend to make those strings for the user or for the developer (in this case, the string of the error is the response sent back to Slack), so I want them to be descriptive and adaptable, testing the string itself with a regular expression or something becomes more rigid. I have also been known to add other handling properties like codes, if I want the exception to change HTTP behavior, or something like that, so the classes give you more options. 

This might also be good for conversation on Monday.